crates/hash/src/lib.rs

```diff
@@ -8,8 +8,19 @@ use essential_types::{convert::bytes_from_word, ContentAddress, Hash, Word};
 use serde::Serialize;
 use sha2::Digest;
 
+mod address_impl;
+pub mod block_addr;
 pub mod contract_addr;
 
+/// Standardized trait for creating content addresses for
+/// types using the correct constructors.
+pub trait Address: Serialize {
+    /// Produce the content address for self.
+    fn content_address(&self) -> ContentAddress {
+        ContentAddress(hash(&self))
+    }
+}
```

Comment:
I'd probably remove the `Serialize` bound and the default impl, then add the implementations manually for `Solution` and `Predicate`, as serialization is not strictly related.
```suggestion
/// Standardized trait for creating content addresses for
/// types using the correct constructors.
pub trait Address {
    /// Produce the content address for self.
    fn content_address(&self) -> ContentAddress;
}
```

Reply:
Isn't `Serialize` always required by `essential_hash::hash()`? Are you saying someone might want to implement this using some other underlying hash scheme that has some custom serialization (i.e. not serde)? If so do we really want to be that flexible?